doc/design/programming_solver.md

```diff
@@ -0,0 +1,280 @@
+[USING glpk]
+INTO result_table;
+```
+
+In the SQL statement:
+
+- `TO MAXIMIZE|MINIMIZE ...` defines an expression string that describes the objective. 
+    - The syntax `MAXIMIZE|MINIMIZE` is used to specify the objective sense. 
+    - In the expression, `SUM` means sum the value across all rows like normal SQL statements.
+- `CONSTRAINT ...` expression strings that describe the constraints, can have multiple `CONSTRAINT` clause lines.
+- `WITH` attributes:
+    - variables: **required**, specify one column that stores the variable name. Using comma to separate if there are multiple variables.
+    - var_type: **required**, specify the variable type for each variable, format like "column:type,column:type...",  the type can be `Integers`, `NonNegativeIntegers`, `Reals` etc.
+- `USING`: **optional**, solver tool to use, default: glpk.
+- `INTO result_table`: set the result table name.
+
+After the SQL statement finishes execution, the result table `result_table` should look like:
+
+| product | result |
+| ------  | ------ |
+| soldier | 20     |
+| train   | 60     |
+
+### Two Variables
+
+For a more general example that have two variables (plants and markets, see the example described in https://en.wikipedia.org/wiki/AMPL for details), we want to minimize the cost of shipment between plants and markets, we have three tables looks like below:
+
+1. Plants capacity table:
+
+    | plants  | capacity |
+    | ------- | -------- |
+    | plantA  | 100      |
+    | plantB  | 90       |
+
+2. Markets demand table:
+
+    | markets |  demand |
+    | ------- | ------- |
+    | marketA | 130     |
+    | marketB | 60      |
+
+3. Plants to markets distance table:
+
+    | plants  | markets | distance |
+    | ------- | ------- | -------- |
+    | plantA  | marketA |  140     |
+    | plantA  | marketB |  210     |
+    | plantB  | marketA |  300     |
+    | plantB  | marketB |  90      |
+
+4. When we start to solve the problem, we'd like to join the tables beforehand:
+
+    ```sql
+    SELECT src.plants, src.markets, src.distance, plants.capacity, markets.demand FROM train_table AS src
+    LEFT JOIN plants ON src.plants = plants.plants
+    LEFT JOIN markets ON src.markets = markets.markets;
+    ```
+    Then we have a "joined" table like below to start the solving process:
+
+    | plants  | markets | distance | capacity | demand |
+    | ------- | ------- | -------- | -------- | ------ |
+    | plantA  | marketA |  140     | 100      | 130    |
+    | plantA  | marketB |  210     | 100      | 60     |
+    | plantB  | marketA |  300     | 90       | 130    |
+    | plantB  | marketB |  90      | 90       | 60     |
+
+Then we can use below extended SQL syntax to describe above example:
+
+```sql
+SELECT src.plants, src.markets, src.distance, plants.capacity, markets.demand FROM train_table AS src
+LEFT JOIN plants ON src.plants = plants.plants
+LEFT JOIN markets ON src.markets = markets.markets
+TO MINIMIZE SUM(distance * 90 / 1000)
```

Comment:
Again, in this problem, the structure of parameters (table `plant_market`) and results (table `result_table`) have the same structure -- the column `result` of `result_table` could be appended to the table `plant_market`.

So, I am wondering could we simplify the SQL statement by 

- removing `WITH variables=..., var_type=...`, but having
- `WITH result` and the column `result` is created with SQL data type INT.